vhp/assembly/Vhp.ts

```diff
@@ -0,0 +1,203 @@
+    result.value = Constants.DECIMALS;
+    return result;
+  }
+
+  total_supply(args?: token.total_supply_arguments): token.total_supply_result {
+    let result = new token.total_supply_result();
+
+    let supplyObject = System.getObject<Uint8Array, token.balance_object>(State.Space.SUPPLY, Constants.SUPPLY_KEY, token.balance_object.decode);
+
+    if (!supplyObject) {
+      result.value = 0;
+    }
+    else {
+      result.value = supplyObject.value;
+    }
+
+    return result;
+  }
+
+  balance_of(args: token.balance_of_arguments): token.balance_of_result {
+    let result = new token.balance_of_result();
+
+    System.require(args.owner != null, 'owner cannot be null');
+
+    let fromBalanceObj = System.getObject<Uint8Array, token.balance_object>(State.Space.BALANCE, args.owner!, token.balance_object.decode);
+
+    if (!fromBalanceObj) {
+      result.value = 0;
+    }
+    else {
+      result.value = fromBalanceObj.value;
+    }
+
+    return result;
+  }
+
+  transfer(args: token.transfer_arguments): token.transfer_result {
+    System.require(args.to != null, 'to cannot be null');
+    System.require(args.from != null, 'from cannot be null');
+    System.require(args.to != args.from, 'cannot transfer to yourself');
+
+    let callerData = System.getCaller();
+    System.require(callerData.caller == args.from || System.checkAuthority(authority.authorization_type.contract_call, args.from!), 'from has not authorized transfer');
+
+    let fromBalanceObj = System.getObject<Uint8Array, token.balance_object>(State.Space.BALANCE, args.from!, token.balance_object.decode);
+
+    if (!fromBalanceObj) {
+      fromBalanceObj = new token.balance_object();
+      fromBalanceObj.value = 0;
+    }
+
+    System.require(fromBalanceObj.value >= args.value, "account 'from' has insufficient balance");
+
+    let toBalanceObj = System.getObject<Uint8Array, token.balance_object>(State.Space.BALANCE, args.to!, token.balance_object.decode);
+
+    if (!toBalanceObj) {
+      toBalanceObj = new token.balance_object();
+      toBalanceObj.value = 0;
+    }
+
+    fromBalanceObj.value -= args.value;
+    toBalanceObj.value += args.value;
+
+    System.putObject(State.Space.BALANCE, args.from!, fromBalanceObj, token.balance_object.encode);
+    System.putObject(State.Space.BALANCE, args.to!, toBalanceObj, token.balance_object.encode);
+
+    let event = new token.transfer_event();
+    event.from = args.from;
+    event.to = args.to;
+    event.value = args.value;
+
+    let impacted: Uint8Array[] = [];
+    impacted.push(args.to!);
+    impacted.push(args.from!);
+
+    System.event('vhp.transfer', Protobuf.encode(event, token.transfer_event.encode), impacted);
+
+    return new token.transfer_result();
+  }
+
+  mint(args: token.mint_arguments): token.mint_result {
+    System.require(args.to != null, "mint argument 'to' cannot be null");
+    System.require(args.value != 0, "mint argument 'value' cannot be zero");
+    if (BUILD_FOR_TESTING) {
```

Comment:
These checks are not quite equivalent to what currently exists in the KOIN contract. Not sure if that is purposeful or not.

``` c++
   const auto [ caller, privilege ] = system::get_caller();
   if ( privilege != chain::privilege::kernel_mode )
   {
#ifdef BUILD_FOR_TESTING
      if( !system::check_authority( constants::contract_id ) )
         system::revert();
#else
      system::log( "Can only mint token from kernel context" );
      return res;
#endif
   }
```

The difference is that kernel mode authorizes whether we are testing or not. In the VHP contract it only authorizes when we are not testing. Which means when we integrate with the PoB contract, it will not be able to mint VHP for burned KOIN unless BUILD_FOR_TESTING is false.